bindings/python/bergamot.cpp

```diff
@@ -0,0 +1,227 @@
+  }
+
+  std::vector<Response> translate(Model model, std::vector<std::string> &inputs,
+                                  const ResponseOptions &options) {
+    py::scoped_ostream_redirect outstream(
+        std::cout,                                // std::ostream&
+        py::module_::import("sys").attr("stdout") // Python output
+    );
+    py::scoped_ostream_redirect errstream(
+        std::cerr,                                // std::ostream&
+        py::module_::import("sys").attr("stderr") // Python output
+    );
+
+    py::call_guard<py::gil_scoped_release> gil_guard;
+
+    // Prepare promises, save respective futures. Have callback's in async set
+    // value to the promises.
+    std::vector<std::future<Response>> futures;
+    std::vector<std::promise<Response>> promises;
+    promises.resize(inputs.size());
+
+    for (size_t i = 0; i < inputs.size(); i++) {
+      auto callback = [&promises, i](Response &&response) {
+        promises[i].set_value(std::move(response));
+      };
+
+      service_.translate(model, std::move(inputs[i]), std::move(callback),
+                         options);
+
+      futures.push_back(std::move(promises[i].get_future()));
+    }
+
+    // Wait on all futures to be ready.
+    std::vector<Response> responses;
+    for (size_t i = 0; i < futures.size(); i++) {
+      futures[i].wait();
+      responses.push_back(std::move(futures[i].get()));
+    }
+
+    return responses;
+  }
+
+  std::vector<Response> pivot(Model first, Model second,
+                              std::vector<std::string> &inputs,
+                              const ResponseOptions &options) {
+    // Prepare promises, save respective futures. Have callback's in async set
+    // value to the promises.
+    std::vector<std::future<Response>> futures;
+    std::vector<std::promise<Response>> promises;
+    promises.resize(inputs.size());
+
+    for (size_t i = 0; i < inputs.size(); i++) {
+      auto callback = [&promises, i](Response &&response) {
+        promises[i].set_value(std::move(response));
+      };
+
+      service_.pivot(first, second, std::move(inputs[i]), std::move(callback),
+                     options);
+
+      futures.push_back(std::move(promises[i].get_future()));
+    }
+
+    // Wait on all futures to be ready.
+    std::vector<Response> responses;
+    for (size_t i = 0; i < futures.size(); i++) {
+      futures[i].wait();
+      responses.push_back(std::move(futures[i].get()));
+    }
+
+    return responses;
+  }
+
+private /*functions*/:
+  static Service make_service(const Service::Config &config) {
+    py::scoped_ostream_redirect outstream(
+        std::cout,                                // std::ostream&
+        py::module_::import("sys").attr("stdout") // Python output
+    );
+    py::scoped_ostream_redirect errstream(
+        std::cerr,                                // std::ostream&
+        py::module_::import("sys").attr("stderr") // Python output
+    );
```

Comment:
Is this because the `Service` constructor could write error messages to either of those?
(Also, how does this work when you release the GIL? If writing to those streams should be handled by python, writing could cause python code to be executed. Does python itself then capture the GIL again?

Reply:
Yes. The factory-method is to place the object on stack (earlier I did it with a `unique_ptr`, then figured I could get away with a factory method for scoped GIL release, redirect and back acquire at end of lifetime during stack unwinding based on scope.

The initial loading messages happen as below and we want that in python stderr or stdout. 


<details>
<summary>Construction output</summary>

```
 [jerin@eltbk lemonade]$ bergamot translate -m en-de-tiny --html=1 --log-level=info <<< 'The quick brown <i id="fox">fox</i> jumps <u>over</u> the lazy <i id="dog">dog</i>.' # Forward translation
[2022-01-19 15:08:58] [data] Loading SentencePiece vocabulary from file /home/jerin/.local/share/lemonade/models/ende.student.tiny11/vocab.deen.spm
[2022-01-19 15:08:59] Missing list of protected prefixes for sentence splitting. Set with --ssplit-prefix-file.
[2022-01-19 15:08:59] [data] Loading binary shortlist as /home/jerin/.local/share/lemonade/models/ende.student.tiny11/lex.s2t.bin false
[2022-01-19 15:08:59] [data] Lexical short list firstNum 50 and bestNum 50
[2022-01-19 15:08:59] [memory] Extending reserved space to 128 MB (device cpu0)
[2022-01-19 15:08:59] Loaded model config
[2022-01-19 15:08:59] Loading scorer of type transformer as feature F0
[2022-01-19 15:08:59] [memory] Reserving 31 MB, device cpu0
[2022-01-19 15:08:59] [memory] Reserving 8 MB, device cpu0
Der schnelle braune <i id="fox">Fuchs</i> springt <u>über</u> den faulen <i id="dog">Hund.</i>
```
</details>

> Does python itself then capture the GIL again?

Scope and C++ RAII. When the objects lifetime's over I'm guessing the destructor is configured to clean-up.

Reply:
Found my answer: [the stream redirect acquires the GIL when syncing to Python's internal buffers](https://github.com/pybind/pybind11/blob/dedda228f44a64d73afff120fd6e47b2a27170c2/include/pybind11/iostream.h#L101)